sysl2/sysl/seqs.go

```diff
@@ -0,0 +1,1081 @@
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/anz-bank/sysl/src/proto"
+	"github.com/sirupsen/logrus"
+)
+
+type indentWriter struct {
+	bytes.Buffer
+	ind      int
+	complete bool // if the current position is a complete line
+}
+
+func (w *indentWriter) Write(p []byte) (n int, err error) {
+	newline := []byte("\n")
+	newlines := bytes.Count(p, newline)
+	if newlines == 0 {
+		if w.complete {
+			w.writeIndent()
+		}
+		n, err = w.Buffer.Write(p)
+		w.complete = false
+		return n, err
+	}
+
+	frags := bytes.SplitN(p, newline, newlines+1)
+
+	for i, frag := range frags {
+		if w.complete {
+			w.writeIndent()
+		}
+		nn, err := w.Buffer.Write(frag)
+		n += nn
+		if err != nil {
+			return n, err
+		}
+		if i+1 < len(frags) {
+			if _, err := w.Buffer.Write(newline); err != nil {
+				return n, err
+			}
+			n++
+		}
+	}
+	w.complete = len(frags[len(frags)-1]) == 0
+	return n, nil
+}
+
+func (w *indentWriter) WriteString(s string) (n int, err error) {
+	if !strings.Contains(s, "\n") {
+		if w.complete {
+			w.writeIndent()
+		}
+		w.complete = false
+		return w.Buffer.WriteString(s)
+	}
+	return w.Write([]byte(s))
+}
+
+func (w *indentWriter) writeIndent() error {
+	if !w.complete {
+		return nil
+	}
+
+	for i := 0; i < w.ind; i++ {
+		if err := w.Buffer.WriteByte(' '); err != nil {
+			return err
+		}
+	}
+
+	w.complete = false
+
+	return nil
+}
+
+func (w *indentWriter) indent() { w.ind++ }
+
+func (w *indentWriter) unindent() {
+	if w.ind == 0 {
+		logrus.Warn("seqs: indentWriter unindented too far")
+		return
+	}
+	w.ind--
+}
+
+type sequenceDiagWriter struct {
+	autogenHeader bool
+	head          indentWriter
+	body          indentWriter
+}
+
+func (w *sequenceDiagWriter) String() string {
+	var sb strings.Builder
+	if w.autogenHeader {
+		fmt.Fprintln(&sb, "''''''''''''''''''''''''''''''''''''''''''")
+		fmt.Fprintln(&sb, "''                                      ''")
+		fmt.Fprintln(&sb, "''  AUTOGENERATED CODE -- DO NOT EDIT!  ''")
+		fmt.Fprintln(&sb, "''                                      ''")
+		fmt.Fprintln(&sb, "''''''''''''''''''''''''''''''''''''''''''")
+		fmt.Fprintln(&sb)
+	}
+
+	fmt.Fprintln(&sb, "@startuml")
+	sb.WriteString(w.head.Buffer.String())
+	sb.WriteString(w.body.Buffer.String())
+	fmt.Fprintln(&sb, "@enduml")
+
+	return sb.String()
+}
+
+type SimpleParser struct {
+	self string
+}
+
+type epFmtParam struct {
+	epname, human, human_sender, needs_int, args, patterns, controls string
```

Comment:
nit: Only bundle same-type fields together if they represent the same kind of intent. I'm not sure how to explain the distinction clearly, but an example might be:
```go
type Vec3 struct {
	x, y, z float32
	length float32
}
```

A good baseline check is, "Does it make sense to compare these things (`==`, `<`, …)?" If the answer is no, then don't put them together. If the answer is yes (as it is for `x` and `length`) it's a judgment call.

An even better heuristic is, "Can I write a single comment describing these things collectively?"
```go
type Vec3 struct {
	x, y, z float32  // Coordinates
	length float32   // Pythagorean length
}
```

(Sidenote: The rule is different for functions, which should favour compactness in order to remain on a single line as often as practicable. However, once the need to go multiline arises, the thinking should be the same.)